### PR TITLE
Penalize images that have a rating count of less then 5 and images th…

### PR DESCRIPTION
**Pull request by p0psicles**
…at are not withing the highest category of resolution, if there are multiple.

- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read the [contribution guide](https://github.com/pymedusa/Medusa/blob/master/.github/CONTRIBUTING.md)

@medariox @labrys

### CODE REVIEW

File: medusa/indexers/indexer_base.py [hunk 1]
```diff
@@ -214,6 +214,26 @@ def _save_images(self, series_id, images):
                 res: resolution such as `1024x768`, `original`, etc
                 id: the image id
         """
+        def penalize_images(images):
+            """Penalize images that have a rating count of less then 5 and images that are not withing the highest
+            category of resolution, if there are multiple."""
+            images = list(images)  # We don't need to the generator.
```
Comment (labrys) [inline medusa/indexers/indexer_base.py]: remove this line

File: medusa/indexers/indexer_base.py [hunk 2]
```diff
@@ -214,6 +214,26 @@ def _save_images(self, series_id, images):
...
+            banner_types = set([image['bannertype2'] for image in images])
+
+            # Try to sort the banner types based on quality (might not be needed)
+            banner_types = sorted(banner_types)
+
+            for image in images:
+                if image['ratingcount'] is not None and image['ratingcount'] < 5:
+                    image['rating'] -= 1
```
Comment (labrys) [inline medusa/indexers/indexer_base.py]: I don't like this type of rating.  Personally i'd go with a weighted average.

File: medusa/indexers/indexer_base.py (hunk 1 above)
Comment (p0psicles) [reply medusa/indexers/indexer_base.py]: can't remove that line. As then after getting the set of bannertype2 the decorator has been used.
`banner_types = set([image['bannertype2'] for image in images])`

And I need to go through it twice.

File: medusa/indexers/indexer_base.py (hunk 2 above)
Comment (p0psicles) [reply medusa/indexers/indexer_base.py]: I understand that's up for debate.

**Comment by labrys** [comment]
Code now uses a weighting system to rate the images by number of votes and rating.  The more votes that make up the rating, the more accurate it considers that rating.  Instead of hard-coded 5 vote minimum, weighting is set to prefer results with more than the average number of votes (specifically more than one standard deviation above the average).

**Comment by labrys** [comment]
Resolution is not considered at this time and will be addressed in future commits.

**Comment by p0psicles** [comment]
@labrys as we discussed in PM, any update on this?

**Comment by labrys** [comment]
Going out of town for a funeral this weekend.  Doubt I'll get a chance to work on it before I get back.

File: medusa/indexers/indexer_base.py [hunk 3]
```diff
@@ -230,8 +230,8 @@ def result(item, threshold, mean):
                 threshold = 1  # Prevent division by zero
             value = item['rating']
             weight = item['ratingcount']
-            res = item['resolution']
-            score = weights.bayesian(weight, value, threshold, mean)
+            res = int(item['resolution'].split('x')[0]) * int(item['resolution'].split('x')[1])
+            score = weights.bayesian(weight * res, value, threshold, mean)
```
Comment (labrys) [inline medusa/indexers/indexer_base.py]: This is exactly why I quit.  Tired of explaining to deaf ears why code like this is bad.

Comment (p0psicles) [reply medusa/indexers/indexer_base.py]: Jip that's me. Mister deaf ears. But at least I finish what i start.

Comment (labrys) [reply medusa/indexers/indexer_base.py]: Easy to finish garbage.

File: medusa/indexers/indexer_base.py [hunk 4]
```diff
@@ -205,6 +209,158 @@ def _set_item(self, sid, seas, ep, attrib, value):  # pylint: disable=too-many-a
...
+
+        def format_result(item):
+            """Format result row for logging output."""
+            row = '{score:>10.3f} {rating:>10.3f} {votes:>6} {res:>15}\t{url}'
+            return row.format(
+                score=item[0],
+                rating=item[1],
+                votes=item[2],
+                res=item[3]['resolution'],
+                url=item[3]['_bannerpath'],
```
Comment (medariox) [inline medusa/indexers/indexer_base.py]: Same index for `resolution` and `_bannerpath`.

Comment (p0psicles) [reply medusa/indexers/indexer_base.py]: yes the fourth index is the image itself.

**Pull request merged**